chartfx-dataset/src/main/java/de/gsi/dataset/spi/DataSetBuilder.java

```diff
@@ -96,72 +114,188 @@ public DataSet build() {
         return dataSet;
     }
 
-    protected DataSet buildRawDataSet(final String dsName) {
-        if (xValues == null && yValues == null) {
-            // no X/Y arrays provided
-            return new DefaultDataSet(dsName, Math.max(initialCapacity, 0));
+    private int getResultDimension() {
+        int maxDim = values.keySet().stream().collect(Collectors.maxBy(Integer::compare)).orElse(-1);
+        maxDim = Math.max(maxDim, errorsNeg.keySet().stream().collect(Collectors.maxBy(Integer::compare)).orElse(-1));
+        maxDim = Math.max(maxDim, errorsPos.keySet().stream().collect(Collectors.maxBy(Integer::compare)).orElse(-1));
+        maxDim = Math.max(maxDim,
+                axisDescriptions.keySet().stream().collect(Collectors.maxBy(Integer::compare)).orElse(-1));
+        if (this.dimension == -1) {
+            return maxDim + 1;
+        } else if (this.dimension <= maxDim) {
+            throw new UnsupportedOperationException("Supplied data dimensions exceed requested number of dimensions");
         }
+        return this.dimension;
+    }
 
-        if (xValues == null) {
-            // no X array provided
-            return buildWithYArrayOnly(dsName);
+    private int[] getResultSize(int nDims) {
+        int[] result = new int[nDims];
+        for (int i = 0; i < nDims; i++) {
+            result[i] = getResultSize(nDims, i);
         }
+        return result;
+    }
 
-        if (yValues == null) {
-            throw new IllegalStateException("yValues is null");
+    private int getResultSize(final int nDims, final int dimIndex) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'getResultSize(int, int)' has an NPath complexity of 640](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=5481451)